File: controle-gastos-web/src/service/categoriaService.ts
import api from "../api/api";
import type { Categoria } from "../models/Categoria";

export const categoriaService = {
  listar: async (): Promise<Categoria[]> => {
    const response = await api.get("/Categorias/Listar");
    return response.data;
  },

  criar: async (descricao: string, finalidade: number): Promise<Categoria> => {
    const response = await api.post("/Categorias/Incluir", { descricao, finalidade });
    return response.data;
  },

  editar: async (id: number, descricao: string, finalidade: number): Promise<Categoria> => {
    const response = await api.put(`/Categorias/Editar/${id}`, { descricao, finalidade });
    return response.data;
  }
};

File: ControleGastos.Api/Services/CategoriaService.cs
using ControleGastos.Api.Models;
using ControleGastos.Api.Repositories;

namespace ControleGastos.Api.Services;

public class CategoriaService
{
    private readonly JsonRepository<Categoria> _repo;

    public CategoriaService()
    {
        _repo = new JsonRepository<Categoria>("categorias.json");
    }

    public List<Categoria> Listar() => _repo.GetAll();

    public async Task <Categoria> Criar(Categoria categoria)
    {
        var categorias = _repo.GetAll();

        categoria.Id = categorias.Any() ? categorias.Max(c => c.Id) + 1 : 1;

        categorias.Add(categoria);
        _repo.SaveAll(categorias);

        return categoria;
    }

    // Editar categoria existente
    public async Task <Categoria> Editar(int id, Categoria categoriaAtualizada)
    {
        var categorias = _repo.GetAll();

        var categoria = categorias.FirstOrDefault(c => c.Id == id);
        if (categoria == null)
            throw new Exception("Categoria não encontrada");

        categoria.Descricao = categoriaAtualizada.Descricao;
        categoria.Finalidade = categoriaAtualizada.Finalidade;

        _repo.SaveAll(categorias);
        return categoria;
    }
}